util/src/KubernetesClient.Util/Cache/Cache.cs

```diff
@@ -0,0 +1,429 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using k8s.Models;
+
+namespace k8s.Util.Cache
+{
+  /// <summary>
+  /// Cache is a C# port of Java's Cache which is a port of k/client-go's ThreadSafeStore. It basically saves and indexes all the entries.
+  /// </summary>
+  /// <typeparam name="TApiType">The type of K8s object to save</typeparam>
+  public class Cache<TApiType> : IIndexer<TApiType>
```

Comment:
We have `IKubernetesObject` in this library which every API type should implement, we should use that qualified here.